### PR TITLE
14167 oauth frontend flow

### LINKED ISSUE #14167: Oauth flow: alternative for 100% frontend applications
Implement the implicit grant mechanism. https://tools.ietf.org/html/rfc6749#section-4.2

### PR DESCRIPTION
**Pull request by javitonino**
Closes #14167 

Implements OAuth implicit flow. This flow has `response_type=token` and returns the token directly in the response (no intermediate authorization code). Further more, redirect URIs use fragment parameters instead of query parameters. Additionally, the endpoints have been modified for CORS support.

Refactors:
- Move `OauthProvider::Strategies` to `OauthProvider::GrantStrategies`
- Create `OauthProvider::TokenStrategies`. Extract current code from controller to the `CodeStrategy`.
- Move `me_public` controller to a different file so it's easier to apply CORS settings.
- Extracted response from OAuth (returning an access token) to a presenter

Changes:
- New `TokenStrategy` which implements the implicit flow details.
- Added some settings to warden to avoid setting cookies for api key authentication
- Added CORS support for public APIs at `carto/api/public/application_controller.rb`

### CODE REVIEW

File: lib/carto/oauth_provider/response_strategies.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        include Scopes
+
+        def self.build_redirect_uri(base_redirect_uri, parameters)
+          redirect_uri = Addressable::URI.parse(base_redirect_uri)
+          redirect_uri.fragment = URI.encode_www_form(parameters)
+
+          redirect_uri.to_s
+        end
+
+        def self.authorize!(oauth_app_user, redirect_uri:, scopes:, state:)
```
[the remark this answers is not in the document]
Comment (javitonino) [reply lib/carto/oauth_provider/response_strategies.rb]: It's fine, we need it for compatibility with the other strategy

**Comment by alonsogarciapablo** [comment]
From [here](https://oauth.net/2/grant-types/implicit/):

>  the industry best practice has changed to recommend that public clients should use either the authorization code flow without the client secret, or use the PKCE extension instead

Out of curiosity: why are we not going with authorization code flow without the client secret? From what I understand, in order for that to work we would need to register apps with an empty `client_secret` so those apps would only support the implicit flow... is this the reason? Thanks!

**Comment by javitonino** [comment]
No. The reason is that the one you link is the only place where it suggests going for the authorization code flow. From reading through the linked documentation:
- It seems to suggest using PKCE for native apps (it doesn't add any/much value for web applications). Basically, PKCE was engineered for this use case (where apps can register handlers for local URLs).
- Authorization code without secret vs implicit flow does not offer a difference in terms of security. However:
  - It introduces an extra round trip for the client, which is not great.
  - Means opening the token endpoint for CORS, which I'd rather keep locked.
  - More opportunities for people to insert the secret in web apps without the need.
  - Harder to restrict `offline` scope for non-backend applications

Upon much reading, I think the paragraph you quote refers mainly to native apps. Also, using implicit flow has much better library support (there is basically no support for PKCE in JS libraries). My conclusion is to follow what Auth0, Google and Facebook are doing. This is the best summary that I could find https://auth0.com/docs/api-auth/which-oauth-flow-to-use.

**Comment by alonsogarciapablo** [comment]
okay okay. Thanks for the explanation!

**Comment by alrocar** [comment]
✅ From the browser: Authorize with `response_type=token` and user logged in returns `#access_token=arA1yL2nH14wh7H5tIZagQ&token_type=Bearer&expires_in=3599.942762863&user_info_url=https%3A%2F%2Foauthtoken.carto-staging.com%2Fapi%2Fv4%2Fme&state=wtf`
✅ From the browser: Authorize with `response_type=token` and user not logged in, redirects to log in, and after log in returns `#access_token=arA1yL2nH14wh7H5tIZagQ&token_type=Bearer&expires_in=3599.942762863&user_info_url=https%3A%2F%2Foauthtoken.carto-staging.com%2Fapi%2Fv4%2Fme&state=wtf`
❓ From JS: Authorize with `response_type=token` returns a CORS error. I've tried both from local and a python http server.

Error:

```
Failed to load https://oauthtoken.carto-staging.com/oauth2/authorize?client_id=K_rFH9V_vfXp&state=wtf&response_type=token: Redirect from 'https://oauthtoken.carto-staging.com/oauth2/authorize?client_id=K_rFH9V_vfXp&state=wtf&response_type=token' to 'https://oauthtoken.carto-staging.com/login' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8000' is therefore not allowed access. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

Code snippet:

```html
<html>

<body>
    <script>
    let link = "https://oauthtoken.carto-staging.com/oauth2/authorize?client_id=K_rFH9V_vfXp&state=wtf&response_type=token"
    fetch(link, {
            mode: "cors",
            redirect: "follow"
        })
        .then(res => res.json())
        .then(response => {
            console.log(response)
        })
        .catch(err => {
            console.log("u")
            console.log(err)
        });
    </script>
</body>

</html>
```

@javitonino Not sure if I'm missing anything

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -45,11 +45,14 @@ def self.ssl_allowed(*_splat)
...
-  Warden::Manager.after_authentication do |user, auth, _opts|
+  Warden::Manager.after_authentication do |user, auth, opts|
     auth.cookies.permanent[ME_ENDPOINT_COOKIE] = {
       value: CartoDB.base_url(user.username),
       domain: Cartodb.config[:session_domain]
-    }
+    } if opts[:store]
+
+    # Do not even send the Set-Cookie header if the strategy did not store anything in the session
+    auth.request.session_options[:skip] = true if opts[:store] == false
```
[the remark this answers is not in the document]
Comment (javitonino) [reply app/controllers/application_controller.rb]: Warden internals. It comes from the `store?`function at the warden strategy (which we define in the `warden.rb` initializer).

**Comment by javitonino** [comment]
@alrocar You should redirect to the authorize endpoint, not use it as an API (because it's not). It is intended to be blocked for JS. If you want to automatically do this, the technique is to use a hidden iframe. e.g: https://willengler.com/ember-oauth2-2/#2performingtheredirectinahiddeniframe

**Comment by alrocar** [comment]
Upss you're right 🤦‍♂️ 

Then moving to deploy 🚀

**Pull request merged**